waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,605 @@
+   This operation will leave Waiter in a sane state on a failure."
+  [{:keys [api-server-url http-client] :as scheduler} service]
+  (when-not service
+    (log/error "Null service passed to kubernetes delete-service.")
+    (ss/throw+ {:status 404 :message "Service not found"}))
+  (let [replicaset-url (str api-server-url
+                            "/apis/extensions/v1beta1/namespaces/"
+                            (:namespace service)
+                            "/replicasets/"
+                            (:k8s-name service))]
+    (patch-object-json replicaset-url http-client
+                       [{:op :replace :path "/metadata/annotations/waiter~1app-status" :value "killed"}
+                        {:op :replace :path "/spec/replicas" :value 0}])
+    (doseq [pod (get-replicaset-pods scheduler service)
+            :let [pod-url (->> pod :metadata :selfLink (str api-server-url))]]
+      (api-request http-client pod-url :request-method :delete))
+    (api-request http-client replicaset-url :request-method :delete)
+    {:message (str "Kubernetes deleted ReplicaSet for " (:k8s-name service))
+     :result :deleted}))
+
+(defn service-id->service
+  "Look up the Kubernetes ReplicaSet associated with a given Waiter service-id,
+   and return a corresponding Waiter Service record."
+  [{:keys [api-server-url http-client service-id->service-description-fn] :as scheduler} service-id]
+  (try
+    (when-let [service-ns (-> service-id service-id->service-description-fn service-description->namespace)]
+      (let [replicasets (->> (str api-server-url
+                                  "/apis/extensions/v1beta1/namespaces/"
+                                  service-ns
+                                  "/replicasets?labelSelector=managed-by=waiter,app="
+                                  (service-id->k8s-name scheduler service-id))
+                             (api-request http-client)
+                             :items)]
+        (when (seq replicasets)
+          (when (second replicasets)
+            (log/warn "Multiple matches found for Waiter Service"
+                      service-id replicasets))
+          (-> replicasets first replicaset->Service))))
+    (catch Throwable e
+      (log/error e "Error creating service record for service-id" service-id)
+      (comment "Returning nil on failure."))))
+
+; The Waiter Scheduler protocol implementation for Kubernetes
+(defrecord KubernetesScheduler [api-server-url http-client
+                                max-conflict-retries
+                                max-name-length
+                                pod-base-port
+                                rs-spec-file-path
+                                service-id->failed-instances-transient-store
+                                service-id->service-description-fn]
+  scheduler/ServiceScheduler
+
+  (get-apps->instances [this]
+    (->> this
+         get-services
+         (pc/map-from-keys (partial instances-breakdown this))))
```

Comment:
Nitpick: I find this more readable:
`(pc/map-from-keys #(instances-breakdown this %) (get-services this))`

Reply:
Done.